server/automated_tests_server.rb

```diff
@@ -43,15 +43,21 @@ def self.perform(markus_address, user_api_key, server_api_key, test_username, te
       start_time = Time.now
       Open3.popen3(run_command, pgroup: true) do |stdin, stdout, stderr, thread|
         pid = thread.pid
-        # mimic capture3 to read safely
-        stdout_thread = Thread.new { stdout.read }
-        stderr_thread = Thread.new { stderr.read }
+        stdin.close
+        # mimic capture3 to read safely and capture each line as it comes so that
+        # these threads don't hang or raise an error if we have to kill them early
+        stdout_thread = Thread.new { stdout.each { |line| output << line + $/} }
+        stderr_thread = Thread.new { stderr.each { |line| errors << line + $/} }
```

Comment:
This is a bit cryptic (and problematic if the record separator is redefined for some reason), I think I prefer the plain old \n.
Also, it's more readable (and usually faster) to use string interpolation rather than concatenation: output << "#{line}\n" (+ don't forget the space before the curly braces)